Update migrations setup during Python unit tests

symbolist opened the pull request:
This PR makes the following changes to database setup during Python unit tests:

1. By default migrations are disabled and the database tables are created from the models. This keeps the developer workflow fast.

2. Passing `--enable-migrations` to `test_system` enables migrations. This can be used on a CI to make sure the migrations are working fine.

In the later case, after the migrations have been run, we delete all the data added by data migrations. This ensures a) whether migrations are run or not, the tests are run from an empty database and b) the unit tests see only the data they setup themselves (otherwise we get lots of test failures).

symbolist commented:
@benpatterson @jimabramson @macdiesel This is finally ready for review and will allow us to run all migrations on Jenkins. Also it would be great if someone else can test this locally. 

@benpatterson FYI It does add a few minutes to the Python shards.

symbolist commented:
When testing locally, checkout the first commit in the branch. The second commit force enables the migrations so that I could test them on Jenkins.

benpatterson commented on openedx/core/djangolib/nose.py:
```diff
@@ -0,0 +1,31 @@
+        # Delete all data added by data migrations. Unit tests should setup their own data using factories.
+        call_command('flush', verbosity=0, interactive=False, load_initial_data=False)
+
+        # Up till Django 1.8, auto increment sequences are not reset when calling flush on a SQLite db.
```
Since we're on django 1.8 already, is this step necessary?

symbolist replied:
No, this is needed on Django 1.8 too, hence the "Up till".

benpatterson commented on pavelib/tests.py:
```diff
@@ -134,6 +138,9 @@ def test_lib(options):
     make_option("-v", "--verbosity", action="count", dest="verbosity", default=1),
     make_option("--pdb", action="store_true", help="Drop into debugger on failures or errors"),
+    make_option(
+        "--enable-migrations", action="store_true", dest="enable_migrations", help="Run migrations instead of syncdb"
```
help text should also point out that this can be set with an environment variable.

andy-armstrong commented on common/djangoapps/util/db.py:
```diff
@@ -231,3 +231,16 @@ def generate_int_id(minimum=0, maximum=MYSQL_MAX_INT, used_ids=None):
     return cid
+
+
+class MigrationModules(object):
```
Nit: this should be named to make it clear what it is doing. Maybe ``NoOpMigrationModules`` or something? The current name sounds like this is a full implementation, not a testing-only one.

andy-armstrong commented on cms/envs/test.py:
```diff
@@ -129,9 +130,11 @@
-MIGRATION_MODULES = {app: "app.migrations_not_used_in_tests" for app in INSTALLED_APPS}
+
+# Disables migrations and creates tables directly from models for speed.
+os.environ['ENABLE_MIGRATIONS'] = '1'
```
I'm confused by this logic. Why is it setting ENABLE_MIGRATIONS to 1, when the comment says that it is disabling migrations? Doesn't the test then always fail because the setting is true?

symbolist replied:
@andy-armstrong Sorry, I should have pointed out that this will be removed. I only added it in a separate commit because I wanted to test the migrations on Jenkins. Once this PR is merged, we will update the Jenkins config to set this variable instead.

andy-armstrong replied:
Ah, got it!

andy-armstrong commented:
:+1: Looks fantastic. Thanks so much for doing this, as debugging tests has been so painful since the 1.8 upgrade.

jimabramson commented:
:+1: :+1: thanks for this improvement @symbolist

benpatterson commented:
Just saw this: "Once this PR is merged, we will update the Jenkins config to set this variable instead."

That should be _part_ of this PR. We can't merge something that then needs to wait for some config change on Jenkins....PRs (and master) will fail in the meantime.

Can we please get the Jenkins config sorted out in this PR as well? That would be done in the scripts (like all-tests.sh)

benpatterson commented:
Actually, in thinking about this, basically devstack will have a different default than jenkins. I'm concerned now that we will get a lot of noise of "worked on my machine..jenkins is doing something special." That puts extra burden on my team, creates confusion, and (worst) creates distrust of the build environment. So, really, I'd like a solution such that, whatever we do on devstack by default, we do on Jenkins by default. I agree this solution is cleaner, but I would rather have a hardcoded exceptions list, or for the default to always run migrations....so that all systems are running tests in the same way by default.

What's the best way to discuss from here?

symbolist commented:
jenkins run bokchoy

symbolist commented:
@benpatterson To be clear, whether we leave the migrations disabled or enable them, all Python unit tests should pass both in devstack and on Jenkins. For example, I have removed the commit which was force enabling the migrations on Jenkins and we have a green build.

Thanks for the pointer to the script. I thought it was in the configuration repo and so needed a PR there. I'll make a change to it to set the `ENABLE_MIGRATIONS` variable.

The only difference between the two options is how the database is set up. In one, we create it directly from the models and in the other we create it from the migrations files. However, in both cases, before the tests are run, the database has the exact same tables which are all empty (at least this is what the intention of this PR is).

symbolist commented:
jenkins run all

benpatterson commented:
Hi @symbolist thanks for the follow-up. 

I do understand the differences between the two, but I mainly want to ensure that a developer needs to actively go against the default behavior. I know there functionally shouldn't b a difference, but one corrupt migration on a branch could cause unnecessary confusion, so I'd like to avoid any unnecessary confusion.

jimabramson commented:
my $.02 (assuming I understand the intent fully):

* before this PR we *always* skipped SOME migrations, in devstack and in regular jenkins builds.
* after this PR, we *always* skip ALL migrations, in devstack and regular jenkins builds.

* before this PR we had no way to enable/test ALL migrations.
* after this PR, we do have that, but it has to be explicitly enabled (either in devstack or a particular jenkins environment).

I think these are both beneficial changes, without a net difference in terms of workflow or risk of unexpected behavior, so I just wanna go on the record as being strongly in favor :)

benpatterson commented:
ok it's possible I'm confusing my PRs @jimabramson @symbolist 

I'll just need another chance for me or Jay to review if that's alright.

symbolist commented:
@jimabramson I have also made the change to run all migrations on Jenkins in this PR now. So once it is merged when running tests locally, sync db will be used to create all tables and on Jenkins migrations will be used to create all tables.

@benpatterson I have updated the help text as you suggested and set the `ENABLE_MIGRATIONS=1` in `scripts/all-tests.sh`. Let me know it is not in the right place. 

I have also incorporated @andy-armstrong's much nicer name suggestion.

benpatterson commented:
Wait I'm sorry...now I'm confused @symbolist . Are you saying that, with the latest change, Jenkins and local dev will work differently by default? That's the part I was trying to avoid...Apologies if I misunderstand.

symbolist commented:
@benpatterson This PR at the least fixes the broken migrations setup and gives us two correctly working options. As far as the defaults are concerned I'll leave it to TestEng to decide what they should be. But I think we should wait and see if the differences in defaults between the two environments do end up causing issues before giving up on a fast developer experience. If at anytime they do become burdensome, only a bool needs to be flipped.

symbolist commented:
jenkins run bokhoy

symbolist commented:
@benpatterson Just saw your last comment. Is that not what you asked in the comments above, to enable migrations on Jenkins?

symbolist commented:
jenkins run bokchoy

benpatterson commented:
@symbolist I know I'm changing my tune a bit, but after discussing with a few folks, it became apparent to me that this could turn into the "works in devstack, but is broken on Jenkins. It must be Jenkins fault so could TestEng please look." That is a huge timesuck for us and rarely results in "Jenkins is wrong". Still we end up spending cycles to prove that it's not the build system. 

So, I'm sorry I'm switching positions on you. I'd really prefer that, by default, Jenkins and development run tests in the same way. I have no problem with giving development a method/env var/etc to operate in an optimized way. (I do this all the time with --fasttest.)

Please let me know when the defaults match and I can give the PR a closer look. And, apologies for the back and forth on my end. I really needed to think it through.

rlucioni commented:
I've updated this branch to address @benpatterson's feedback as part of my hackathon project. By default, migrations continue to be applied as they always have been. Exporting DISABLE_MIGRATIONS=1 or passing --disable-migrations to Paver commands will create tables directly from apps' models.

benpatterson commented:
thanks @rlucioni and @symbolist for making sure the defaults on devstack and jenkins match.

:+1: from me once python tests pass on Jenkins.

rlucioni commented:
jenkins run bokchoy

jzoldak commented:
@rlucioni or @symbolist please update testing.rst with an example of what devs should do to run tests locally with this turned on.

rlucioni commented:
@jzoldak I will follow up with a PR updating testing.rst.

symbolist commented:
Thank @rlucioni for the cleanup. I was working on adding some tests for migrations but I'll put that into a separate PR.
